[Blog] A Deep-Dive into Flink's Network Stack

NicoK opened the pull request:
Submitting the markup file for the first blog post in a series of posts about Flink's network stack.

Also, please note that while rebuilding the web site it appeared that the current status was not clean with respect to the `feed.xml` which I fixed in the first commit.

rmetzger commented on _posts/2019-06-04-flink-network-stack.md:
```diff
@@ -0,0 +1,324 @@
+
+* Transport:
+    * **high throughput:**
+    Sending each record one-by-one obviously does not give the best throughput. For this, Flink has to buffer a bunch of records into its network buffers and send them altogether.
```
```suggestion
    Instead of sending each record one-by-one, Flink buffers a bunch of records into its network buffers and send them altogether. This reduces the cost per record and leads to high throughput.
```

rmetzger replied:
Just a suggestion :)

rmetzger commented on _posts/2019-06-04-flink-network-stack.md:
```diff
@@ -0,0 +1,324 @@
+<br>
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2019-06-04-network-stack/flink-network-stack3.png" width="700px" alt="Physical-transport-backpressure-Flink's Network Stack"/>
```
Is B.4 is causing the backpressure?

NicoK replied:
I clarified this a bit better now / added a description of what the picture is showing

rmetzger commented on _posts/2019-06-04-flink-network-stack.md:
```diff
@@ -0,0 +1,324 @@
+<center>
+<img src="{{ site.baseurl }}/img/blog/2019-06-04-network-stack/flink-network-stack9.png" width="650px" alt="Network-buffertimeout-Flink's Network Stack"/>
+</center>
+<br>
```
it might make sense to write down the results from the benchmark as well.
People might not get what the picture wants to tell them. (and I see a risk of people just scrolling through the pictures, drawing conclusions themselves)

rmetzger commented:
Very well written, pretty detailed.
I think it is good to go.

If you want to further improve it: I really like the technical depth of the text, but I sometimes got lost at the details.
By assuming less knowledge by the readers (thus giving more background information or explicitly stating the obvious), we might open up this information to more readers (at the cost of making the text longer).
One question I had while reading: Are there metrics available for the network stack to look at?

NicoK commented:
Thanks for the review @MarkSfik and @rmetzger. I addressed the comments you had and added one or two more headings and paragraphs to make things a bit clearer. Please have a look whether it is better now.

MarkSfik commented on _posts/2019-06-04-flink-network-stack.md:
```diff
@@ -0,0 +1,337 @@
+
+## Inflicting Backpressure (2)
+
+As opposed to the receiver's backpressure mechanisms without flow control, credits provide a more direct control: If a receiver cannot keep up, its available credits will eventually hit 0 and stop the sender from forwarding buffers to the lower network stack. There is backpressure on this logical channel only and no need to block reading from a multiplexed TCP channel. Other receivers are therefo ...
```
```suggestion
As opposed to the receiver's backpressure mechanisms without flow control, credits provide a more direct control: If a receiver cannot keep up, its available credits will eventually hit 0 and stop the sender from forwarding buffers to the lower network stack. There is backpressure on this logical channel only and there is no need to block reading from a multiplexed TCP channel. Other receivers are therefore not affected in processing available buffers.
```

MarkSfik commented on _posts/2019-06-04-flink-network-stack.md:
```diff
@@ -0,0 +1,337 @@
+
+<img align="right" src="{{ site.baseurl }}/img/blog/2019-06-04-network-stack/flink-network-stack5.png" width="300" height="200" alt="Physical-transport-credit-flow-checkpoints-Flink's Network Stack"/>
+
+Since, with flow control, a channel in a multiplex can not block another of its logical channels, the overall resource utilisation should increase. In addition, by having full control over how much data is “on the wire”, we are also able to improve [checkpoint alignments](https://ci.apache.org/projects/flink/flink-docs-release-1.8/internals/stream_checkpointing.html#checkpointing): without flow c ...
```
```suggestion
Since, with flow control, a channel in a multiplex cannot block another of its logical channels, the overall resource utilisation should increase. In addition, by having full control over how much data is “on the wire”, we are also able to improve [checkpoint alignments](https://ci.apache.org/projects/flink/flink-docs-release-1.8/internals/stream_checkpointing.html#checkpointing): without flow control, it would take a while for the channel to fill the network stack’s internal buffers and propagate that the receiver is not reading anymore. During that time, a lot of buffers could be sitting around. Any checkpoint barrier would have to queue up behind these buffers and would thus have to wait until all of those have been processed before it can start (“Barriers never overtake records!”).
```

MarkSfik commented:
Thanks for adding the extra parts @NicoK I left a couple of comments, but other than that it looks good to me!

NicoK commented:
thanks, @MarkSfik, I integrated your proposed changes

NicoK commented:
Thanks @rmetzger. Merging now...
